hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/TimeoutOptimizer.java

```diff
@@ -0,0 +1,227 @@
+        }
+        else if (opType == AbfsRestOperationType.SetAcl) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_SET_ACL_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.SetOwner) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_SET_OWNER_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.SetPermissions) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_SET_PERMISSIONS_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.Append) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_APPEND_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.CheckAccess) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_CHECK_ACCESS_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.GetPathStatus) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_GET_PATH_STATUS_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.Flush) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_FLUSH_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.ReadFile) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_READFILE_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.LeasePath) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_LEASE_PATH_REQUEST_TIMEOUT);
+        }
+        if (timeout == null) {
+            timeout = DEFAULT_TIMEOUT;
+        }
+        requestTimeout = Integer.parseInt(timeout);
+        readTimeout = requestTimeout;
+        connTimeout = requestTimeout - 1;
+        updateUrl();
+    }
+
+    private void updateTimeouts(int retryCount) {
+        if (retryCount == 0) {
+            return;
+        }
+        int maxRetryCount = retryPolicy.getRetryCount();
+        if (retryCount <= maxRetryCount && timeoutIncRate > 0) {
+            // retry count is still valid
+            // timeout increment rate is a valid value
+            if ((requestTimeout * timeoutIncRate) > maxReqTimeout) {
+                requestTimeout = maxReqTimeout;
+            } else {
+                requestTimeout *= timeoutIncRate;
+            }
+            readTimeout = requestTimeout;
+            connTimeout = requestTimeout - 1;
+        }
+    }
+
+    private void updateUrl() {
+        // updates URL with existing request timeout value
+        URL updatedUrl = null;
+        try {
+            URIBuilder uriBuilder = new URIBuilder(url.toURI());
+            uriBuilder.setParameter(HttpQueryParams.QUERY_PARAM_TIMEOUT, Integer.toString(requestTimeout));
+            updatedUrl = uriBuilder.build().toURL();
+        } catch (URISyntaxException e) {
+
+        } catch (MalformedURLException e) {
+
+        }
```

Comment:
should RuntimeException be thrown here?

Reply:
should we throw back the exception in catch block ?

Reply:
Had kept this empty because an already formed URL will be used (just modifying one query parameter), but can throw RuntimeException over here.

Reply:
i understand that url will always be correct on which we would do manipulation, but still better to throw RuntimeException. maybe in future, some other class wants to use this timeoutOptimizer, it can use it. This class should be agnostic to whatever flowing in.

Reply:
Have added RuntimeException for both the catch blocks in TimeoutOptimizer class.